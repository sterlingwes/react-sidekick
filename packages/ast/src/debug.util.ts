const debug = process.env.DEBUG;
export const log = (...args: any[]) => {
  if (debug) console.log(...args);
};
