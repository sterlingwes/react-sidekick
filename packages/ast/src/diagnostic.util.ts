import { Node, SyntaxKind } from "typescript";

export enum DiagnosticReason {
  SavedElement,
  Interesting,
  Skipped,
  Unknown,
}

export interface DiagnosticTree {
  id?: string;
  kind?: SyntaxKind;
  reason?: DiagnosticReason;
  children: DiagnosticTree[];
}

export const trackDiagnostic = (
  reason: DiagnosticReason,
  kind: SyntaxKind,
  tree?: DiagnosticTree,
  id?: string
) => {
  if (!tree) return;

  const newNode = {
    id,
    kind: kind,
    reason,
    children: [],
  };

  tree.children.push(newNode);

  return newNode;
};

export const trackDeadEndDiagnostic = (node: Node, tree?: DiagnosticTree) => {
  if (!tree) return;

  const newNode = {
    kind: node.kind,
    reason: DiagnosticReason.Skipped,
    children: [],
  };

  tree.children.push(newNode);

  node.forEachChild((child) => trackDeadEndDiagnostic(child, newNode));

  return newNode;
};

export const trackDiagnosticFileBoundary = (
  filePath: string,
  tree?: DiagnosticTree
) => {
  if (!tree) return;

  const newNode = {
    id: filePath,
    children: [],
  };

  tree.children.push(newNode);

  return newNode;
};
