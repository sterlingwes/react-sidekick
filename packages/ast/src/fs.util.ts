import fs from "fs";
import { CrawlPaths } from "./types";

export const interestingCrawlPaths = (
  crawlPaths: CrawlPaths,
  componentIdentifiers: Set<string>
) => {
  return Object.keys(crawlPaths).filter((path) => {
    if (/\./.test(path) === false) {
      // skip non-relative paths
      return false;
    }

    const bindings = crawlPaths[path];
    return bindings.some((binding) => componentIdentifiers.has(binding));
  });
};

const hasExtension = (filePath: string) => /\.[a-z0-9A-Z]{2,3}/.test(filePath);
const extensions = [".ts", ".tsx", "/index.ts", "/index.tsx"];

const fileExists = (path: string) => {
  return new Promise((resolve) => {
    fs.access(path, (err) => (err ? resolve(false) : resolve(true)));
  });
};

const asyncFindExtensionMatch = (extensions: string[], filePath: string) => {
  return extensions.reduce((chain, ext) => {
    return chain.then((match) => {
      if (match) return match;
      const testPath = `${filePath}${ext}`;
      return fileExists(testPath).then((exists) =>
        exists ? testPath : undefined
      );
    });
  }, Promise.resolve() as Promise<undefined | string>);
};

export const findPath = async (filePath: string) => {
  if (hasExtension(filePath)) {
    return filePath;
  }

  return asyncFindExtensionMatch(extensions, filePath);
};
