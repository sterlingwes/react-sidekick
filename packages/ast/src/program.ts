import path from "path";
import { createProject } from "@ts-morph/bootstrap";
import { readdir, readFile } from "fs";
import ts, { JsxEmit } from "typescript";
import { CompilerOptions } from "./types";

const readFilePromise = (path: string) =>
  new Promise<string>((resolve, reject) => {
    readFile(path, (err, data) =>
      err ? reject(err) : resolve(data.toString())
    );
  });

const directoryOfInterest = (dirName: string) =>
  dirName !== "node_modules" && !dirName.startsWith("__");
const fileOfInterest = (fileName: string) => fileName.endsWith(".tsx");

const resolveFilesFromDirectory = (
  dirPath: string,
  filesAccumulated: string[] = []
): Promise<string[]> => {
  return new Promise((resolve, reject) => {
    readdir(dirPath, { withFileTypes: true }, (err, dirents) => {
      if (err) return reject(err);
      const files = dirents.reduce((chain, item) => {
        return chain.then((acc) => {
          if (item.isFile() && fileOfInterest(item.name)) {
            return acc.concat(`${dirPath}/${item.name}`);
          }

          if (item.isDirectory() && directoryOfInterest(item.name)) {
            return resolveFilesFromDirectory(`${dirPath}/${item.name}`).then(
              (recursedFiles) => acc.concat(recursedFiles)
            );
          }

          return acc;
        });
      }, Promise.resolve(filesAccumulated) as Promise<string[]>);

      resolve(files);
    });
  });
};

const buildVirtualProgram = async (rootEntryFilePath: string) => {
  const project = await createProject();
  const rootDir = path.dirname(rootEntryFilePath);
  const files = await resolveFilesFromDirectory(rootDir);
  await files.reduce((chain, filePath) => {
    return chain
      .then(() => readFilePromise(filePath))
      .then((fileContents) => {
        project.createSourceFile(filePath, fileContents);
      });
  }, Promise.resolve() as Promise<unknown>);

  return project.createProgram();
};

export const buildProgram = async (
  projectRootEntryFilePath: string,
  compilerOptions?: CompilerOptions
) => {
  if (compilerOptions?.virtualFs) {
    const program = await buildVirtualProgram(projectRootEntryFilePath);
    return program;
  }

  return ts.createProgram([projectRootEntryFilePath], {
    noEmitOnError: true,
    noImplicitAny: false,
    target: ts.ScriptTarget.ES5,
    module: ts.ModuleKind.CommonJS,
    jsx: JsxEmit.React,
  });
};
