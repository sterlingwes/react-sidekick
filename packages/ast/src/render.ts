import { SyntaxKind } from "typescript";
import { DiagnosticReason, DiagnosticTree } from "./diagnostic.util";
import { tab } from "./string.util";
import { NodeLookups, NodeTree } from "./types";

export const renderTreeText = (tree: NodeTree, depth = 1): string => {
  const { name } = tree;
  if (tree.children.length === 0) {
    return `<${name} />`;
  }
  return `<${name}>\n${
    tree.children
      .map((child) => tab(depth) + renderTreeText(child, depth + 1) + "\n")
      .join("") + tab(depth - 1)
  }</${name}>`;
};

export const renderDiagnosticText = (
  tree: DiagnosticTree,
  depth = 1
): string => {
  const { id, reason, kind } = tree;
  let name = "????????";

  if (id && !reason && !kind) {
    name = `!---- ${id} ----------------------------`;
  } else if (kind && reason) {
    name = `${SyntaxKind[kind]} ${DiagnosticReason[reason]}`;
  } else {
    name = `${SyntaxKind[kind ?? SyntaxKind.Unknown]}${id ? ` (${id}) ` : " "}${
      DiagnosticReason[reason ?? DiagnosticReason.Unknown]
    }`;
  }

  if (tree.children.length === 0) {
    return `<${name} />`;
  }
  return `<${name}>\n${
    tree.children
      .map(
        (child) => tab(depth) + renderDiagnosticText(child, depth + 1) + "\n"
      )
      .join("") + tab(depth - 1)
  }</${name}>`;
};
