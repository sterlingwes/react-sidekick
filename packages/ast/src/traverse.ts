import ts, {
  Identifier,
  ImportDeclaration,
  ImportSpecifier,
  JsxElement,
  JsxOpeningElement,
  JsxSelfClosingElement,
  Node,
  SourceFile,
  SyntaxKind,
} from "typescript";
import { log } from "./debug.util";
import {
  createNode,
  encodeId,
  interesting,
  jsxValueName,
  target,
} from "./node.util";
import { AstState, NodeLookups, NodeTree } from "./types";

const saveJsxElement = (
  node: JsxElement | JsxSelfClosingElement,
  tree: NodeTree,
  lookups: NodeLookups,
  path: number[]
) => {
  let element: JsxSelfClosingElement | JsxOpeningElement;

  if (node.kind === SyntaxKind.JsxElement) {
    element = node.openingElement;
  } else {
    element = node;
  }

  let name = (element.tagName as Identifier).escapedText as string;
  const { tagName, attributes } = element;

  switch (tagName.kind) {
    case SyntaxKind.Identifier:
      name = tagName.escapedText as string;
      break;
    case SyntaxKind.PropertyAccessExpression:
      name = (tagName.expression as Identifier).escapedText as string;
      name += ".";
      name += tagName.name.escapedText as string;
      break;
  }

  const props = attributes.properties.reduce((acc, attr) => {
    if (attr.kind !== SyntaxKind.JsxAttribute) {
      return acc;
    }

    return {
      ...acc,
      [attr.name.escapedText as string]: jsxValueName(attr.initializer),
    };
  }, {});

  const id = encodeId(name, path);
  const savedNode = createNode(id);
  lookups.elements[id] = { name };
  tree.children.push(savedNode);
  return savedNode;
};

let lastIdentifier: string | undefined;

const traverse = (
  node: Node | SourceFile | undefined,
  tree: NodeTree,
  lookups: NodeLookups,
  path: number[]
) => {
  if (!node) {
    log("! no node ! did you pass a valid entry file path?");
    process.exit(1);
  }

  let filteredIndex = 0;
  node.forEachChild((childNode) => {
    // save name of component export
    if (childNode.kind === SyntaxKind.Identifier) {
      lastIdentifier = (childNode as Identifier).escapedText as string;
    }

    if ([SyntaxKind.ImportDeclaration].includes(childNode.kind)) {
      const importNode = childNode as ImportDeclaration;
      let bindings: any[] = [];
      importNode.importClause?.namedBindings?.forEachChild((binding) => {
        bindings.push(binding);
      });

      if (bindings.length === 0) {
        const singleImport = importNode.importClause;
        bindings.push(singleImport);
      }

      log("import", {
        bindings: bindings.map(
          (binding: ImportSpecifier) => binding.name.escapedText
        ),
        module: (importNode.moduleSpecifier as any).text,
      });
    }

    if (target(childNode)) {
      let parentNode = tree;

      if (lastIdentifier) {
        path.push(filteredIndex);
        filteredIndex = 0; // reset index for depth change
        const parentNodeId = encodeId(lastIdentifier, path);
        parentNode = createNode(parentNodeId);
        lookups.elements[parentNodeId] = { name: lastIdentifier };
        tree.children.push(parentNode);
        lastIdentifier = undefined;
      }
      const newPath = [...path, filteredIndex];
      const newTree = saveJsxElement(childNode, parentNode, lookups, newPath);
      traverse(childNode, newTree, lookups, newPath);
      filteredIndex++;
    } else if (interesting(childNode)) {
      traverse(childNode, tree, lookups, path);
    }
  });
};

export const traverseFromFile = (filePath: string): AstState => {
  const program = ts.createProgram([filePath], {
    noEmitOnError: true,
    noImplicitAny: false,
    target: ts.ScriptTarget.ES5,
    module: ts.ModuleKind.CommonJS,
  });

  const sourceFile = program.getSourceFile(filePath);

  const nodeTree = createNode("_root");
  const lookups = {
    files: {},
    elements: {},
  };

  traverse(sourceFile, nodeTree, lookups, []);

  return {
    ...lookups,
    hierarchy: nodeTree,
  };
};

const repeat = (str: string, times: number) =>
  Array.from(new Array(times)).reduce((acc) => acc + str, "");

const tab = (times: number) => repeat("  ", times);

export const renderTreeText = (
  tree: NodeTree,
  elements: NodeLookups["elements"],
  depth = 1
): string => {
  const el = elements[tree.id];
  const name = el ? el.name : tree.id;
  if (tree.children.length === 0) {
    return `<${name} />`;
  }
  return `<${name}>\n${
    tree.children
      .map(
        (child) =>
          tab(depth) + renderTreeText(child, elements, depth + 1) + "\n"
      )
      .join("") + tab(depth - 1)
  }</${name}>`;
};
