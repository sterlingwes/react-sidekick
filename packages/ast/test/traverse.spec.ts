import { DiagnosticTree } from "../src/diagnostic.util";
import { buildProgram } from "../src/program";
import { renderDiagnosticText, renderTreeText } from "../src/render";
import { traverseFromFile, traverseProject } from "../src/traverse";

describe("sample-app tests", () => {
  const dirname = "./packages/ast/test/sample-app";
  const entryPath = "./packages/ast/test/sample-app/Main.tsx";
  let state: Awaited<ReturnType<typeof traverseFromFile>>;

  beforeEach(async () => {
    state = await traverseProject(entryPath, {
      plugins: [require("../src/libraries/react-navigation")],
    });
  });

  it("should represent the full component hierarchy", () => {
    const renderedAst = renderTreeText(state.hierarchy);
    expect(renderedAst).toMatchInlineSnapshot(`
      "<_Root>
        <Main>
          <RRProvider>
            <NavigationContainer>
              <Stack.Navigator>
                <MenuButton />
                <Stack.Screen>
                  <ActionList>
                    <ActionList>
                      <TouchableOpacity>
                        <View>
                          <Text />
                        </View>
                      </TouchableOpacity>
                    </ActionList>
                  </ActionList>
                </Stack.Screen>
                <Stack.Screen>
                  <DetailScreen>
                    <DetailScreen>
                      <View>
                        <Text />
                        <Text />
                      </View>
                    </DetailScreen>
                  </DetailScreen>
                </Stack.Screen>
                <Stack.Group>
                  <Stack.Screen>
                    <SettingsScreen>
                      <SettingsScreen>
                        <SafeAreaView>
                          <Pressable>
                            <View>
                              <Text />
                            </View>
                          </Pressable>
                          <Pressable>
                            <View>
                              <Text />
                            </View>
                          </Pressable>
                        </SafeAreaView>
                      </SettingsScreen>
                    </SettingsScreen>
                  </Stack.Screen>
                  <Stack.Screen>
                    <SettingsProfileScreen>
                      <SettingsProfileScreen>
                        <View>
                          <Text />
                          <TextInput />
                          <TextInput />
                          <Text />
                        </View>
                      </SettingsProfileScreen>
                    </SettingsProfileScreen>
                  </Stack.Screen>
                </Stack.Group>
              </Stack.Navigator>
            </NavigationContainer>
          </RRProvider>
        </Main>
      </_Root>"
    `);
  });

  describe("files", () => {
    it("should annotate nodes with file Ids used in files lookup", () => {
      expect(state.files).toEqual({
        "1": "packages/ast/test/sample-app/Main.tsx",
        "2": "./packages/ast/test/sample-app/screens/ActionList.tsx",
        "3": "./packages/ast/test/sample-app/screens/DetailScreen.tsx",
        "4": "./packages/ast/test/sample-app/components/index.ts",
        "5": "./packages/ast/test/sample-app/screens/SettingsProfileScreen.tsx",
        "6": "./packages/ast/test/sample-app/screens/SettingsScreen.tsx",
      });
    });
  });

  describe("elements lookup", () => {
    it("should have keys with IDs for each unique component in the hierarchy", () => {
      const keys = Object.keys(state.elements);
      expect(keys).toMatchInlineSnapshot(`
        Array [
          "Main-0",
          "RRProvider-0.0",
          "NavigationContainer-0.0.0",
          "Stack.Navigator-0.0.0.0",
          "MenuButton-0.0.0.0.0",
          "Stack.Screen-0.0.0.0.1",
          "ActionList-0.0.0.0.1.0",
          "Stack.Screen-0.0.0.0.2",
          "DetailScreen-0.0.0.0.2.0",
          "Stack.Group-0.0.0.0.3",
          "Stack.Screen-0.0.0.0.3.0",
          "SettingsScreen-0.0.0.0.3.0.0",
          "Stack.Screen-0.0.0.0.3.1",
          "SettingsProfileScreen-0.0.0.0.3.1.0",
          "ActionList-0.0.0.0.1.0.0",
          "TouchableOpacity-0.0.0.0.1.0.0.0",
          "View-0.0.0.0.1.0.0.0.0",
          "Text-0.0.0.0.1.0.0.0.0.0",
          "View-0.0.0.0.1.0.0.1",
          "View-0.0.0.0.1.0.0.1.0",
          "Text-0.0.0.0.1.0.0.1.0.0",
          "FlatList-0.0.0.0.1.0.0.1.1",
          "RefreshControl-0.0.0.0.1.0.0.1.1.0",
          "DetailScreen-0.0.0.0.2.0.0",
          "View-0.0.0.0.2.0.0.0",
          "Text-0.0.0.0.2.0.0.0.0",
          "Text-0.0.0.0.2.0.0.0.1",
          "SettingsProfileScreen-0.0.0.0.3.1.0.0",
          "View-0.0.0.0.3.1.0.0.0",
          "Text-0.0.0.0.3.1.0.0.0.0",
          "TextInput-0.0.0.0.3.1.0.0.0.1",
          "TextInput-0.0.0.0.3.1.0.0.0.2",
          "Text-0.0.0.0.3.1.0.0.0.3",
          "SettingsScreen-0.0.0.0.3.0.0.0",
          "SafeAreaView-0.0.0.0.3.0.0.0.0",
          "Pressable-0.0.0.0.3.0.0.0.0.0",
          "View-0.0.0.0.3.0.0.0.0.0.0",
          "Text-0.0.0.0.3.0.0.0.0.0.0.0",
          "Pressable-0.0.0.0.3.0.0.0.0.1",
          "View-0.0.0.0.3.0.0.0.0.1.0",
          "Text-0.0.0.0.3.0.0.0.0.1.0.0",
        ]
      `);
    });
  });

  describe("leafNodes lookup", () => {
    beforeEach(async () => {
      const program = await buildProgram(entryPath);
      const sourceFile = program.getSourceFile(entryPath);
      if (!sourceFile) throw new Error("No sourcefile!");

      state = await traverseFromFile(sourceFile, {
        projectFiles: [entryPath],
        program,
        dirname,
      });
    });

    it("should have IDs in the set representing tree leaf nodes (no children)", () => {
      const ids = Array.from(state.leafNodes);
      expect(ids).toMatchInlineSnapshot(`
        Array [
          "MenuButton-0.0.0.0.0",
          "Stack.Screen-0.0.0.0.1",
          "Stack.Screen-0.0.0.0.2",
          "Stack.Screen-0.0.0.0.3.0",
          "Stack.Screen-0.0.0.0.3.1",
        ]
      `);
    });
  });

  describe("virtual fs program", () => {
    it("should complete traversal & build the same hierarchy as non-virtual", async () => {
      const virtualState = await traverseProject(entryPath, {
        plugins: [require("../src/libraries/react-navigation")],
        compilerOptions: {
          virtualFs: true,
        },
      });

      const renderedVirtualState = renderTreeText(virtualState.hierarchy);

      const nonVirtualState = await traverseProject(entryPath, {
        plugins: [require("../src/libraries/react-navigation")],
      });

      const renderedNonVirtualState = renderTreeText(nonVirtualState.hierarchy);

      expect(renderedNonVirtualState).toEqual(renderedVirtualState);
    });
  });

  describe("diagnostics", () => {
    let state: Awaited<ReturnType<typeof traverseProject>>;
    let diagnosticTree: DiagnosticTree;

    beforeEach(async () => {
      diagnosticTree = { children: [] };

      state = await traverseProject(entryPath, {
        runDiagnostic: true,
        plugins: [require("../src/libraries/react-navigation")],
      });
    });

    it("should populate the tree with reasons for node selection", () => {
      if (!state.diagnosticTree) throw new Error("Expected diagnostic result");
      const diagnostic = renderDiagnosticText(state.diagnosticTree);
      expect(diagnostic).toMatchSnapshot();
    });
  });

  describe("plugins", () => {
    describe("react navigation", () => {
      beforeEach(async () => {
        const program = await buildProgram(entryPath);
        const sourceFile = program.getSourceFile(entryPath);
        if (!sourceFile) throw new Error("No sourcefile!");

        state = await traverseFromFile(sourceFile, {
          plugins: [require("../src/libraries/react-navigation")],
          projectFiles: [entryPath],
          program,
          dirname,
        });
      });

      it("should have its own state linking screen route names to component names", () => {
        expect(Object.keys(state.thirdParty)).toEqual(["reactNavigation"]);
        expect((state.thirdParty.reactNavigation as any).routes).toEqual({
          Detail: "DetailScreen",
          Home: "ActionList",
          Settings: "SettingsScreen",
          SettingsProfile: "SettingsProfileScreen",
        });
        expect(
          Array.from((state.thirdParty.reactNavigation as any).components)
        ).toEqual([
          "ActionList",
          "DetailScreen",
          "SettingsScreen",
          "SettingsProfileScreen",
        ]);
      });

      it("should change the leaf nodes in the main AST state", () => {
        const ids = Array.from(state.leafNodes);
        expect(ids).toMatchInlineSnapshot(`
          Array [
            "MenuButton-0.0.0.0.0",
            "ActionList-0.0.0.0.1.0",
            "DetailScreen-0.0.0.0.2.0",
            "SettingsScreen-0.0.0.0.3.0.0",
            "SettingsProfileScreen-0.0.0.0.3.1.0",
          ]
        `);
      });
    });
  });
});
