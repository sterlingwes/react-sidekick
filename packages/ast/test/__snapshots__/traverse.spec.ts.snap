// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sample-app tests diagnostics should populate the tree with reasons for node selection 1`] = `
"<FirstToken Unknown>
  <!---- packages/ast/test/sample-app/Main.tsx ---------------------------->
    <ImportDeclaration Skipped>
      <ImportClause Skipped>
        <Identifier Skipped />
      </ImportClause Skipped>
    </ImportDeclaration Skipped>
    <ImportDeclaration Skipped>
      <ImportClause Skipped>
        <NamedImports Skipped>
          <ImportSpecifier Skipped>
            <Identifier Skipped />
          </ImportSpecifier Skipped>
        </NamedImports Skipped>
      </ImportClause Skipped>
    </ImportDeclaration Skipped>
    <ImportDeclaration Skipped>
      <ImportClause Skipped>
        <NamedImports Skipped>
          <ImportSpecifier Skipped>
            <Identifier Skipped />
          </ImportSpecifier Skipped>
        </NamedImports Skipped>
      </ImportClause Skipped>
    </ImportDeclaration Skipped>
    <ImportDeclaration Skipped>
      <ImportClause Skipped>
        <NamedImports Skipped>
          <ImportSpecifier Skipped>
            <Identifier Skipped />
          </ImportSpecifier Skipped>
        </NamedImports Skipped>
      </ImportClause Skipped>
    </ImportDeclaration Skipped>
    <ImportDeclaration Skipped>
      <ImportClause Skipped>
        <NamedImports Skipped>
          <ImportSpecifier Skipped>
            <Identifier Skipped />
          </ImportSpecifier Skipped>
        </NamedImports Skipped>
      </ImportClause Skipped>
    </ImportDeclaration Skipped>
    <ImportDeclaration Skipped>
      <ImportClause Skipped>
        <NamedImports Skipped>
          <ImportSpecifier Skipped>
            <Identifier Skipped />
          </ImportSpecifier Skipped>
        </NamedImports Skipped>
      </ImportClause Skipped>
    </ImportDeclaration Skipped>
    <ImportDeclaration Skipped>
      <ImportClause Skipped>
        <NamedImports Skipped>
          <ImportSpecifier Skipped>
            <Identifier Skipped />
          </ImportSpecifier Skipped>
        </NamedImports Skipped>
      </ImportClause Skipped>
    </ImportDeclaration Skipped>
    <ImportDeclaration Skipped>
      <ImportClause Skipped>
        <NamedImports Skipped>
          <ImportSpecifier Skipped>
            <Identifier Skipped />
          </ImportSpecifier Skipped>
        </NamedImports Skipped>
      </ImportClause Skipped>
    </ImportDeclaration Skipped>
    <ImportDeclaration Skipped>
      <ImportClause Skipped>
        <NamedImports Skipped>
          <ImportSpecifier Skipped>
            <Identifier Skipped />
          </ImportSpecifier Skipped>
        </NamedImports Skipped>
      </ImportClause Skipped>
    </ImportDeclaration Skipped>
    <ImportDeclaration Skipped>
      <ImportClause Skipped>
        <NamedImports Skipped>
          <ImportSpecifier Skipped>
            <Identifier Skipped />
          </ImportSpecifier Skipped>
        </NamedImports Skipped>
      </ImportClause Skipped>
    </ImportDeclaration Skipped>
    <FirstStatement Interesting>
      <VariableDeclarationList Interesting>
        <VariableDeclaration Interesting>
          <Identifier Skipped />
          <CallExpression Interesting>
            <Identifier Skipped />
          </CallExpression Interesting>
        </VariableDeclaration Interesting>
      </VariableDeclarationList Interesting>
    </FirstStatement Interesting>
    <FirstStatement Interesting>
      <VariableDeclarationList Interesting>
        <VariableDeclaration Interesting>
          <Identifier Skipped />
          <CallExpression Interesting>
            <Identifier Skipped />
          </CallExpression Interesting>
        </VariableDeclaration Interesting>
      </VariableDeclarationList Interesting>
    </FirstStatement Interesting>
    <FirstStatement Interesting>
      <ExportKeyword Skipped />
      <VariableDeclarationList Interesting>
        <VariableDeclaration Interesting>
          <Identifier Skipped />
          <ArrowFunction Interesting>
            <EqualsGreaterThanToken Skipped />
            <Block Interesting>
              <ReturnStatement Interesting>
                <ParenthesizedExpression Interesting>
                  <JsxElement (RRProvider-0.0) SavedElement>
                    <JsxOpeningElement Interesting>
                      <Identifier Skipped />
                      <JsxAttributes Interesting>
                        <JsxAttribute Interesting>
                          <Identifier Skipped />
                          <JsxExpression Interesting>
                            <Identifier Skipped />
                          </JsxExpression Interesting>
                        </JsxAttribute Interesting>
                      </JsxAttributes Interesting>
                    </JsxOpeningElement Interesting>
                    <JsxText Skipped />
                    <JsxElement (NavigationContainer-0.0.0) SavedElement>
                      <JsxOpeningElement Interesting>
                        <Identifier Skipped />
                        <JsxAttributes Interesting />
                      </JsxOpeningElement Interesting>
                      <JsxText Skipped />
                      <JsxElement (Stack.Navigator-0.0.0.0) SavedElement>
                        <JsxOpeningElement Interesting>
                          <PropertyAccessExpression Skipped>
                            <Identifier Skipped />
                          </PropertyAccessExpression Skipped>
                          <JsxAttributes Interesting>
                            <JsxAttribute Interesting>
                              <Identifier Skipped />
                              <JsxExpression Interesting>
                                <ArrowFunction Interesting>
                                  <Parameter Skipped>
                                    <ObjectBindingPattern Skipped>
                                      <BindingElement Skipped>
                                        <Identifier Skipped />
                                      </BindingElement Skipped>
                                    </ObjectBindingPattern Skipped>
                                  </Parameter Skipped>
                                  <EqualsGreaterThanToken Skipped />
                                  <ParenthesizedExpression Interesting>
                                    <ObjectLiteralExpression Interesting>
                                      <PropertyAssignment Interesting>
                                        <Identifier Skipped />
                                        <ArrowFunction Interesting>
                                          <EqualsGreaterThanToken Skipped />
                                          <ParenthesizedExpression Interesting>
                                            <JsxSelfClosingElement (MenuButton-0.0.0.0.0) SavedElement>
                                              <Identifier Skipped />
                                              <JsxAttributes Interesting>
                                                <JsxAttribute Interesting>
                                                  <Identifier Skipped />
                                                  <StringLiteral Skipped />
                                                </JsxAttribute Interesting>
                                                <JsxAttribute Interesting>
                                                  <Identifier Skipped />
                                                  <JsxExpression Interesting>
                                                    <ArrowFunction Interesting>
                                                      <EqualsGreaterThanToken Skipped />
                                                      <CallExpression Interesting>
                                                        <PropertyAccessExpression Skipped>
                                                          <Identifier Skipped />
                                                        </PropertyAccessExpression Skipped>
                                                        <StringLiteral Skipped />
                                                      </CallExpression Interesting>
                                                    </ArrowFunction Interesting>
                                                  </JsxExpression Interesting>
                                                </JsxAttribute Interesting>
                                              </JsxAttributes Interesting>
                                            </JsxSelfClosingElement (MenuButton-0.0.0.0.0) SavedElement>
                                          </ParenthesizedExpression Interesting>
                                        </ArrowFunction Interesting>
                                      </PropertyAssignment Interesting>
                                    </ObjectLiteralExpression Interesting>
                                  </ParenthesizedExpression Interesting>
                                </ArrowFunction Interesting>
                              </JsxExpression Interesting>
                            </JsxAttribute Interesting>
                          </JsxAttributes Interesting>
                        </JsxOpeningElement Interesting>
                        <JsxText Skipped />
                        <JsxSelfClosingElement (Stack.Screen-0.0.0.0.1) SavedElement>
                          <PropertyAccessExpression Skipped>
                            <Identifier Skipped />
                          </PropertyAccessExpression Skipped>
                          <JsxAttributes Interesting>
                            <JsxAttribute Interesting>
                              <Identifier Skipped />
                              <StringLiteral Skipped />
                            </JsxAttribute Interesting>
                            <JsxAttribute Interesting>
                              <Identifier Skipped />
                              <JsxExpression Interesting>
                                <Identifier Skipped />
                              </JsxExpression Interesting>
                            </JsxAttribute Interesting>
                          </JsxAttributes Interesting>
                        </JsxSelfClosingElement (Stack.Screen-0.0.0.0.1) SavedElement>
                        <JsxText Skipped />
                        <JsxSelfClosingElement (Stack.Screen-0.0.0.0.2) SavedElement>
                          <PropertyAccessExpression Skipped>
                            <Identifier Skipped />
                          </PropertyAccessExpression Skipped>
                          <JsxAttributes Interesting>
                            <JsxAttribute Interesting>
                              <Identifier Skipped />
                              <StringLiteral Skipped />
                            </JsxAttribute Interesting>
                            <JsxAttribute Interesting>
                              <Identifier Skipped />
                              <JsxExpression Interesting>
                                <Identifier Skipped />
                              </JsxExpression Interesting>
                            </JsxAttribute Interesting>
                          </JsxAttributes Interesting>
                        </JsxSelfClosingElement (Stack.Screen-0.0.0.0.2) SavedElement>
                        <JsxText Skipped />
                        <JsxElement (Stack.Group-0.0.0.0.3) SavedElement>
                          <JsxOpeningElement Interesting>
                            <PropertyAccessExpression Skipped>
                              <Identifier Skipped />
                            </PropertyAccessExpression Skipped>
                            <JsxAttributes Interesting>
                              <JsxAttribute Interesting>
                                <Identifier Skipped />
                                <JsxExpression Interesting>
                                  <ObjectLiteralExpression Interesting>
                                    <PropertyAssignment Interesting>
                                      <Identifier Skipped />
                                      <StringLiteral Skipped />
                                    </PropertyAssignment Interesting>
                                  </ObjectLiteralExpression Interesting>
                                </JsxExpression Interesting>
                              </JsxAttribute Interesting>
                            </JsxAttributes Interesting>
                          </JsxOpeningElement Interesting>
                          <JsxText Skipped />
                          <JsxSelfClosingElement (Stack.Screen-0.0.0.0.3.0) SavedElement>
                            <PropertyAccessExpression Skipped>
                              <Identifier Skipped />
                            </PropertyAccessExpression Skipped>
                            <JsxAttributes Interesting>
                              <JsxAttribute Interesting>
                                <Identifier Skipped />
                                <StringLiteral Skipped />
                              </JsxAttribute Interesting>
                              <JsxAttribute Interesting>
                                <Identifier Skipped />
                                <JsxExpression Interesting>
                                  <Identifier Skipped />
                                </JsxExpression Interesting>
                              </JsxAttribute Interesting>
                            </JsxAttributes Interesting>
                          </JsxSelfClosingElement (Stack.Screen-0.0.0.0.3.0) SavedElement>
                          <JsxText Skipped />
                          <JsxSelfClosingElement (Stack.Screen-0.0.0.0.3.1) SavedElement>
                            <PropertyAccessExpression Skipped>
                              <Identifier Skipped />
                            </PropertyAccessExpression Skipped>
                            <JsxAttributes Interesting>
                              <JsxAttribute Interesting>
                                <Identifier Skipped />
                                <StringLiteral Skipped />
                              </JsxAttribute Interesting>
                              <JsxAttribute Interesting>
                                <Identifier Skipped />
                                <JsxExpression Interesting>
                                  <Identifier Skipped />
                                </JsxExpression Interesting>
                              </JsxAttribute Interesting>
                            </JsxAttributes Interesting>
                          </JsxSelfClosingElement (Stack.Screen-0.0.0.0.3.1) SavedElement>
                          <JsxText Skipped />
                          <JsxClosingElement Skipped>
                            <PropertyAccessExpression Skipped>
                              <Identifier Skipped />
                            </PropertyAccessExpression Skipped>
                          </JsxClosingElement Skipped>
                        </JsxElement (Stack.Group-0.0.0.0.3) SavedElement>
                        <JsxText Skipped />
                        <JsxClosingElement Skipped>
                          <PropertyAccessExpression Skipped>
                            <Identifier Skipped />
                          </PropertyAccessExpression Skipped>
                        </JsxClosingElement Skipped>
                      </JsxElement (Stack.Navigator-0.0.0.0) SavedElement>
                      <JsxText Skipped />
                      <JsxClosingElement Skipped>
                        <Identifier Skipped />
                      </JsxClosingElement Skipped>
                    </JsxElement (NavigationContainer-0.0.0) SavedElement>
                    <JsxText Skipped />
                    <JsxClosingElement Skipped>
                      <Identifier Skipped />
                    </JsxClosingElement Skipped>
                  </JsxElement (RRProvider-0.0) SavedElement>
                </ParenthesizedExpression Interesting>
              </ReturnStatement Interesting>
            </Block Interesting>
          </ArrowFunction Interesting>
        </VariableDeclaration Interesting>
      </VariableDeclarationList Interesting>
    </FirstStatement Interesting>
    <EndOfFileToken Skipped />
    <!---- /Users/wesjohnson/Code/react-sidekick/packages/ast/test/sample-app/components/index.ts ---------------------------->
      <ExportDeclaration Skipped>
        <StringLiteral Skipped />
      </ExportDeclaration Skipped>
      <EndOfFileToken Skipped />
    </!---- /Users/wesjohnson/Code/react-sidekick/packages/ast/test/sample-app/components/index.ts ---------------------------->
  </!---- packages/ast/test/sample-app/Main.tsx ---------------------------->
</FirstToken Unknown>"
`;
